dev/run_0001/parameters.txt
model.spectral_grid
SpectralGrid:
├ Spectral:   T31 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
├ Grid:       48-ring OctahedralGaussianGrid{Float32}, 3168 grid points
├ Resolution: 401km (average)
├ Vertical:   8-layer atmosphere, 2-layer land
└ Device:     CPU using Array

model.device_setup
SpeedyWeather.DeviceSetup{CPU, DataType}(CPU(), KernelAbstractions.CPU, 4)

model.dynamics
true

model.geometry
Geometry for SpectralGrid:
├ Spectral:   T31 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
├ Grid:       48-ring OctahedralGaussianGrid{Float32}, 3168 grid points
├ Resolution: 401km (average)
├ Vertical:   8-layer atmosphere, 2-layer land
└ Device:     CPU using Array

model.planet
Earth{Float32} <: SpeedyWeather.AbstractPlanet
├ rotation::Float32 = 7.29e-5
├ gravity::Float32 = 9.81
├ daily_cycle::Bool = true
├ length_of_day::Second = 86400 seconds
├ seasonal_cycle::Bool = true
├ length_of_year::Second = 31557600 seconds
├ equinox::DateTime = 2000-03-20T00:00:00
├ axial_tilt::Float32 = 23.4
└ solar_constant::Float32 = 1365.0

model.atmosphere
EarthAtmosphere{Float32} <: SpeedyWeather.AbstractAtmosphere
├ mol_mass_dry_air::Float32 = 28.9649
├ mol_mass_vapour::Float32 = 18.0153
├ heat_capacity::Float32 = 1004.0
├ R_gas::Float32 = 8.3145
├ R_dry::Float32 = 287.05432
├ R_vapour::Float32 = 461.52438
├ mol_ratio::Float32 = 0.62197006
├ μ_virt_temp::Float32 = 0.60779446
├ κ::Float32 = 0.2859107
├ water_density::Float32 = 1000.0
├ latent_heat_condensation::Float32 = 2.501e6
├ latent_heat_sublimation::Float32 = 2.801e6
├ stefan_boltzmann::Float32 = 5.67e-8
├ pres_ref::Float32 = 100000.0
├ temp_ref::Float32 = 288.0
├ moist_lapse_rate::Float32 = 0.005
├ dry_lapse_rate::Float32 = 0.0098
└ layer_thickness::Float32 = 8500.0

model.coriolis
Coriolis{Float32, Vector{Float32}} <: SpeedyWeather.AbstractCoriolis
├ nlat::Int64 = 48
└── arrays: f

model.geopotential
Geopotential{Float32, Vector{Float32}} <: SpeedyWeather.AbstractGeopotential
├ nlayers::Int64 = 8
└── arrays: Δp_geopot_half, Δp_geopot_full

model.adiabatic_conversion
AdiabaticConversion{Float32, Vector{Float32}} <: SpeedyWeather.AbstractAdiabaticConversion
├ nlayers::Int64 = 8
└── arrays: σ_lnp_A, σ_lnp_B

model.particle_advection
NoParticleAdvection <: SpeedyWeather.AbstractParticleAdvection


model.initial_conditions
InitialConditions{ZonalWind, PressureOnOrography, JablonowskiTemperature, ZeroInitially} <: SpeedyWeather.AbstractInitialConditions
├ vordiv::ZonalWind = ZonalWind <: SpeedyWeather.AbstractInitialConditions
├ η₀::Float64 = 0.252
├ u₀::Float64 = 35.0
├ perturb_lat::Float64 = 40.0
├ perturb_lon::Float64 = 20.0
├ perturb_uₚ::Float64 = 1.0
└ perturb_radius::Float64 = 0.1
├ pres::PressureOnOrography = PressureOnOrography <: SpeedyWeather.AbstractInitialConditions

├ temp::JablonowskiTemperature = JablonowskiTemperature <: SpeedyWeather.AbstractInitialConditions
├ η₀::Float64 = 0.252
├ σ_tropopause::Float64 = 0.2
├ u₀::Float64 = 35.0
├ ΔT::Float64 = 0.0
└ Tmin::Float64 = 200.0
└ humid::ZeroInitially = ZeroInitially <: SpeedyWeather.AbstractInitialConditions


model.forcing
NoForcing <: SpeedyWeather.AbstractForcing


model.drag
NoDrag <: SpeedyWeather.AbstractDrag


model.random_process
NoRandomProcess <: SpeedyWeather.AbstractRandomProcess


model.tracers
Dict{Symbol, Tracer}()

model.orography
EarthOrography{Float32, OctahedralGaussianGrid{Float32}, LowerTriangularMatrix{ComplexF32}} <: SpeedyWeather.AbstractOrography
├ path::String = SpeedyWeather.jl/input_data
├ file::String = orography.nc
├ file_Grid::UnionAll = FullGaussianGrid
├ scale::Float32 = 1.0
├ smoothing::Bool = true
├ smoothing_power::Float32 = 1.0
├ smoothing_strength::Float32 = 0.1
├ smoothing_fraction::Float32 = 0.05
└── arrays: orography, geopot_surf

model.land_sea_mask
EarthLandSeaMask{Float32, OctahedralGaussianGrid{Float32}} <: AbstractLandSeaMask
├ path::String = SpeedyWeather.jl/input_data
├ file::String = land-sea_mask.nc
├ file_Grid::UnionAll = FullClenshawGrid
└── arrays: mask

model.ocean
SeasonalOceanClimatology{Float32, OctahedralGaussianArray{Float32, 2, Matrix{Float32}}} <: AbstractOcean
├ nlat_half::Int64 = 24
├ path::String = SpeedyWeather.jl/input_data
├ file::String = sea_surface_temperature.nc
├ varname::String = sst
├ file_Grid::UnionAll = FullGaussianGrid
├ missing_value::Float32 = NaN
└── arrays: monthly_temperature

model.land
DryLandModel <: SpeedyWeather.AbstractDryLand
├ spectral_grid: SpectralGrid
├ geometry: LandGeometry{Float32, Vector{Float32}}
├ thermodynamics: LandThermodynamics{Float32}
└ temperature: SeasonalLandTemperature{Float32, OctahedralGaussianArray{Float32, 2, Matrix{Float32}}}

model.solar_zenith
SolarZenith{Float32} <: AbstractZenith
├ length_of_day::Second = 86400 seconds
├ length_of_year::Second = 31557600 seconds
├ equation_of_time::Bool = true
├ seasonal_cycle::Bool = true
├ solar_declination::SpeedyWeather.SinSolarDeclination{Float32} = SpeedyWeather.SinSolarDeclination{Float32} <: AbstractSolarDeclination
├ axial_tilt::Float32 = 23.4
├ equinox::DateTime = 2000-03-20T00:00:00
├ length_of_year::Second = 31557600 seconds
└ length_of_day::Second = 86400 seconds
├ time_correction::SpeedyWeather.SolarTimeCorrection{Float32} = SpeedyWeather.SolarTimeCorrection{Float32} <: AbstractSolarTimeCorrection
├ a::Float32 = 0.004297
├ s1::Float32 = -1.837877
├ c1::Float32 = 0.107029
├ s2::Float32 = -2.340475
└ c2::Float32 = -0.837378
└ initial_time::Base.RefValue{DateTime} = Base.RefValue{DateTime}(DateTime("2000-01-01T00:00:00"))

model.albedo
Albedo <: SpeedyWeather.AbstractAlbedo
├ ocean: GlobalConstantAlbedo{Float32}
└ land: GlobalConstantAlbedo{Float32}

model.physics
true

model.boundary_layer_drag
BulkRichardsonDrag{Float32} <: SpeedyWeather.AbstractBoundaryLayer
├ κ::Float32 = 0.4
├ z₀::Float32 = 3.21e-5
├ Ri_c::Float32 = 10.0
└ drag_max::Base.RefValue{Float32} = Base.RefValue{Float32}(0.0005774734f0)

model.temperature_relaxation
NoTemperatureRelaxation <: SpeedyWeather.AbstractTemperatureRelaxation


model.vertical_diffusion
BulkRichardsonDiffusion{Float32, Vector{Float32}} <: SpeedyWeather.AbstractVerticalDiffusion
├ nlayers::Int64 = 8
├ κ::Float32 = 0.4
├ z₀::Float32 = 3.21e-5
├ Ri_c::Float32 = 10.0
├ fb::Float32 = 0.1
├ diffuse_static_energy::Bool = true
├ diffuse_momentum::Bool = true
├ diffuse_humidity::Bool = true
├ logZ_z₀::Base.RefValue{Float32} = Base.RefValue{Float32}(16.645391f0)
├ sqrtC_max::Base.RefValue{Float32} = Base.RefValue{Float32}(0.024030676f0)
└── arrays: ∇²_above, ∇²_below

model.surface_thermodynamics
SurfaceThermodynamicsConstant <: SpeedyWeather.AbstractSurfaceThermodynamics


model.surface_wind
SurfaceWind{Float32} <: SpeedyWeather.AbstractSurfaceWind
├ f_wind::Float32 = 0.95
├ V_gust::Float32 = 5.0
├ use_boundary_layer_drag::Bool = true
├ drag_land::Float32 = 0.0024
└ drag_sea::Float32 = 0.0018

model.surface_heat_flux
SurfaceHeatFlux{SurfaceOceanHeatFlux{Float32}, SurfaceLandHeatFlux{Float32}} <: SpeedyWeather.AbstractSurfaceHeatFlux
├ ocean::SurfaceOceanHeatFlux{Float32} = SurfaceOceanHeatFlux{Float32} <: SpeedyWeather.AbstractSurfaceHeatFlux
├ use_boundary_layer_drag::Bool = true
└ heat_exchange::Float32 = 0.0009
└ land::SurfaceLandHeatFlux{Float32} = SurfaceLandHeatFlux{Float32} <: SpeedyWeather.AbstractSurfaceHeatFlux
├ use_boundary_layer_drag::Bool = true
└ heat_exchange::Float32 = 0.0012

model.convection
DryBettsMiller{Float32, NoSurfacePerturbation} <: SpeedyWeather.AbstractConvection
├ nlayers::Int64 = 8
├ time_scale::Second = 14400 seconds
└ surface_temp::NoSurfacePerturbation = NoSurfacePerturbation <: SpeedyWeather.AbstractSurfacePerturbation


model.optical_depth
ZeroOpticalDepth{Float32} <: SpeedyWeather.AbstractOpticalDepth


model.shortwave_radiation
TransparentShortwave <: SpeedyWeather.AbstractShortwave


model.longwave_radiation
JeevanjeeRadiation{Float32} <: SpeedyWeather.AbstractLongwave
├ α::Float32 = 0.025
├ temp_tropopause::Float32 = 200.0
└ time_scale::Second = 86400 seconds

model.stochastic_physics
NoStochasticPhysics <: SpeedyWeather.AbstractStochasticPhysics


model.time_stepping
Leapfrog{Float32} <: SpeedyWeather.AbstractTimeStepper
├ trunc::Int64 = 31
├ nsteps::Int64 = 2
├ Δt_at_T31::Second = 2400 seconds
├ radius::Float32 = 6.371e6
├ adjust_with_output::Bool = true
├ robert_filter::Float32 = 0.1
├ williams_filter::Float32 = 0.53
├ Δt_millisec::Millisecond = 2400000 milliseconds
├ Δt_sec::Float32 = 2400.0
└ Δt::Float32 = 0.00037670694

model.spectral_transform
SpectralTransform{Float32, Array}:
├ Spectral:   T31, 33x32 LowerTriangularMatrix{Complex{Float32}}
├ Grid:       48-ring OctahedralGaussianArray{Float32}
├ Truncation: dealiasing = 2 (quadratic)
├ Legendre:   Polynomials 53.83 KB, shortcut: linear
└ Memory:     for 8 layers (182.53 KB)

model.implicit
ImplicitPrimitiveEquation{Float32, Vector{Float32}, Matrix{Float32}, Array{Float32, 3}} <: SpeedyWeather.AbstractImplicit
├ trunc::Int64 = 31
├ nlayers::Int64 = 8
├ α::Float32 = 1.0
├ ξ::Base.RefValue{Float32} = Base.RefValue{Float32}(0.0f0)
└── arrays: temp_profile, R, U, L, W, L0, L1, L2, L3, L4, S, S⁻¹

model.horizontal_diffusion
HyperDiffusion{Float32, Matrix{Float32}} <: SpeedyWeather.AbstractHorizontalDiffusion
├ trunc::Int64 = 31
├ nlayers::Int64 = 8
├ power::Float32 = 4.0
├ time_scale::Second = 14400 seconds
├ time_scale_div::Second = 3600 seconds
├ resolution_scaling::Float32 = 1.0
├ power_stratosphere::Float32 = 2.0
├ tapering_σ::Float32 = 0.2
└── arrays: expl, impl, expl_div, impl_div

model.vertical_advection
CenteredVerticalAdvection{Float32, 1}()

model.output
NetCDFOutput{FullGaussianGrid{Float32}}
├ status: active
├ write restart file: true (if active)
├ interpolator: AnvilInterpolator{Float32, OctahedralGaussianGrid, SpeedyWeather.RingGrids.GridGeometry{OctahedralGaussianGrid, Vector{Float64}, Vector{Int64}, Vector{UnitRange{Int64}}}, SpeedyWeather.RingGrids.AnvilLocator{Float32, Vector{Float32}, Vector{Int64}}}
├ path: /home/runner/work/SpeedyWeather.jl/SpeedyWeather.jl/docs/build/run_0001/output.nc
├ frequency: 21600 seconds
└┐ variables:
 ├ w: vertical velocity dσ/dt [s^-1]
 ├ v: meridional wind [m/s]
 ├ u: zonal wind [m/s]
 └ vor: relative vorticity [s^-1]

model.callbacks
Dict{Symbol, SpeedyWeather.AbstractCallback}()

model.feedback
Feedback <: AbstractFeedback
├ verbose::Bool = false
├ debug::Bool = true
├ output::Bool = true
├ id::String = 0001
├ run_path::String = /home/runner/work/SpeedyWeather.jl/SpeedyWeather.jl/docs/build/run_0001
├ progress_meter::ProgressMeter.Progress = ProgressMeter.Progress <: ProgressMeter.AbstractProgress
├ n::Int64 = 1
├ start::Int64 = 0
├ barlen::Nothing = nothing
├ barglyphs::ProgressMeter.BarGlyphs = ProgressMeter.BarGlyphs('|', '█', ['▏', '▎', '▍', '▌', '▋', '▊', '▉'], ' ', '|')
└ core::ProgressMeter.ProgressCore = ProgressMeter.ProgressCore(:green, "Weather is speedy: run 0001 ", 0.1, false, 0, IOContext(Base.PipeEndpoint(RawFD(-1) closed, 0 bytes waiting)), true, 1, 0, ReentrantLock(nothing, 0x00000000, 0x00, Base.GenericCondition{Base.Threads.SpinLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), (-4849946800566399402, -4860226519185164219, 4289040937940112555)), 0, 1, false, false, 1.74292896997363e9, 1.74292896997363e9, 1.74292896997363e9)
├ progress_txt::IOStream = IOStream()
└ nars_detected::Bool = false



dev/run_0002/parameters.txt
model.spectral_grid
SpectralGrid:
├ Spectral:   T63 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
├ Grid:       96-ring OctahedralGaussianGrid{Float32}, 10944 grid points
├ Resolution: 216km (average)
├ Vertical:   1-layer atmosphere, 2-layer land
└ Device:     CPU using Array

model.device_setup
SpeedyWeather.DeviceSetup{CPU, DataType}(CPU(), KernelAbstractions.CPU, 4)

model.geometry
Geometry for SpectralGrid:
├ Spectral:   T63 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
├ Grid:       96-ring OctahedralGaussianGrid{Float32}, 10944 grid points
├ Resolution: 216km (average)
├ Vertical:   1-layer atmosphere, 2-layer land
└ Device:     CPU using Array

model.planet
Earth{Float32} <: SpeedyWeather.AbstractPlanet
├ rotation::Float32 = 7.29e-5
├ gravity::Float32 = 9.81
├ daily_cycle::Bool = true
├ length_of_day::Second = 86400 seconds
├ seasonal_cycle::Bool = true
├ length_of_year::Second = 31557600 seconds
├ equinox::DateTime = 2000-03-20T00:00:00
├ axial_tilt::Float32 = 23.4
└ solar_constant::Float32 = 1365.0

model.atmosphere
EarthAtmosphere{Float32} <: SpeedyWeather.AbstractAtmosphere
├ mol_mass_dry_air::Float32 = 28.9649
├ mol_mass_vapour::Float32 = 18.0153
├ heat_capacity::Float32 = 1004.0
├ R_gas::Float32 = 8.3145
├ R_dry::Float32 = 287.05432
├ R_vapour::Float32 = 461.52438
├ mol_ratio::Float32 = 0.62197006
├ μ_virt_temp::Float32 = 0.60779446
├ κ::Float32 = 0.2859107
├ water_density::Float32 = 1000.0
├ latent_heat_condensation::Float32 = 2.501e6
├ latent_heat_sublimation::Float32 = 2.801e6
├ stefan_boltzmann::Float32 = 5.67e-8
├ pres_ref::Float32 = 100000.0
├ temp_ref::Float32 = 288.0
├ moist_lapse_rate::Float32 = 0.005
├ dry_lapse_rate::Float32 = 0.0098
└ layer_thickness::Float32 = 8500.0

model.coriolis
Coriolis{Float32, Vector{Float32}} <: SpeedyWeather.AbstractCoriolis
├ nlat::Int64 = 96
└── arrays: f

model.orography
NoOrography{Float32, OctahedralGaussianGrid{Float32}, LowerTriangularMatrix{ComplexF32}} <: SpeedyWeather.AbstractOrography
└── arrays: orography, geopot_surf

model.forcing
NoForcing <: SpeedyWeather.AbstractForcing


model.drag
NoDrag <: SpeedyWeather.AbstractDrag


model.particle_advection
NoParticleAdvection <: SpeedyWeather.AbstractParticleAdvection


model.initial_conditions
ZonalJet <: SpeedyWeather.AbstractInitialConditions
├ latitude::Float64 = 45.0
├ width::Float64 = 19.28571428571429
├ umax::Float64 = 80.0
├ perturb_lat::Float64 = 45.0
├ perturb_lon::Float64 = 270.0
├ perturb_xwidth::Float64 = 19.098593171027442
├ perturb_ywidth::Float64 = 3.819718634205488
└ perturb_height::Float64 = 120.0

model.random_process
NoRandomProcess <: SpeedyWeather.AbstractRandomProcess


model.tracers
Dict{Symbol, Tracer}()

model.time_stepping
Leapfrog{Float32} <: SpeedyWeather.AbstractTimeStepper
├ trunc::Int64 = 63
├ nsteps::Int64 = 2
├ Δt_at_T31::Second = 2400 seconds
├ radius::Float32 = 6.371e6
├ adjust_with_output::Bool = true
├ robert_filter::Float32 = 0.1
├ williams_filter::Float32 = 0.53
├ Δt_millisec::Millisecond = 1200000 milliseconds
├ Δt_sec::Float32 = 1200.0
└ Δt::Float32 = 0.00018835347

model.spectral_transform
SpectralTransform{Float32, Array}:
├ Spectral:   T63, 65x64 LowerTriangularMatrix{Complex{Float32}}
├ Grid:       96-ring OctahedralGaussianArray{Float32}
├ Truncation: dealiasing = 2 (quadratic)
├ Legendre:   Polynomials 411.72 KB, shortcut: linear
└ Memory:     for 1 layers (82.50 KB)

model.implicit
ImplicitShallowWater{Float32, Vector{Float32}} <: SpeedyWeather.AbstractImplicit
├ trunc::Int64 = 63
├ α::Float32 = 1.0
├ H::Base.RefValue{Float32} = Base.RefValue{Float32}(8500.0f0)
├ ξH::Base.RefValue{Float32} = Base.RefValue{Float32}(3.202009f0)
└── arrays: g∇², ξg∇², S⁻¹

model.horizontal_diffusion
HyperDiffusion{Float32, Matrix{Float32}} <: SpeedyWeather.AbstractHorizontalDiffusion
├ trunc::Int64 = 63
├ nlayers::Int64 = 1
├ power::Float32 = 4.0
├ time_scale::Second = 14400 seconds
├ time_scale_div::Second = 3600 seconds
├ resolution_scaling::Float32 = 1.0
├ power_stratosphere::Float32 = 2.0
├ tapering_σ::Float32 = 0.2
└── arrays: expl, impl, expl_div, impl_div

model.output
NetCDFOutput{FullGaussianGrid{Float32}}
├ status: active
├ write restart file: true (if active)
├ interpolator: AnvilInterpolator{Float32, OctahedralGaussianGrid, SpeedyWeather.RingGrids.GridGeometry{OctahedralGaussianGrid, Vector{Float64}, Vector{Int64}, Vector{UnitRange{Int64}}}, SpeedyWeather.RingGrids.AnvilLocator{Float32, Vector{Float32}, Vector{Int64}}}
├ path: /home/runner/work/SpeedyWeather.jl/SpeedyWeather.jl/docs/build/run_0002/output.nc
├ frequency: 21600 seconds
└┐ variables:
 ├ eta: interface displacement [m]
 ├ v: meridional wind [m/s]
 ├ u: zonal wind [m/s]
 └ vor: relative vorticity [s^-1]

model.callbacks
Dict{Symbol, SpeedyWeather.AbstractCallback}()

model.feedback
Feedback <: AbstractFeedback
├ verbose::Bool = false
├ debug::Bool = true
├ output::Bool = true
├ id::String = 0002
├ run_path::String = /home/runner/work/SpeedyWeather.jl/SpeedyWeather.jl/docs/build/run_0002
├ progress_meter::ProgressMeter.Progress = ProgressMeter.Progress <: ProgressMeter.AbstractProgress
├ n::Int64 = 431
├ start::Int64 = 0
├ barlen::Nothing = nothing
├ barglyphs::ProgressMeter.BarGlyphs = ProgressMeter.BarGlyphs('|', '█', ['▏', '▎', '▍', '▌', '▋', '▊', '▉'], ' ', '|')
└ core::ProgressMeter.ProgressCore = ProgressMeter.ProgressCore(:green, "Weather is speedy: run 0002 ", 0.1, false, 0, IOContext(Base.PipeEndpoint(RawFD(-1) closed, 0 bytes waiting)), true, 1, 431, ReentrantLock(nothing, 0x00000000, 0x00, Base.GenericCondition{Base.Threads.SpinLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), (34, 3465, 1)), 0, 1, false, false, 1.742928989650554e9, 1.742928989650554e9, 1.742928989650554e9)
├ progress_txt::IOStream = IOStream()
└ nars_detected::Bool = false



CHANGELOG.md
# Changelog

## Unreleased


- CUDA-ised spectral transform [#602](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/602)
- Revised extended CI tests for differentiability can be launched on comment in PRs [#695](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/695), [#696](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/696), [#697](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/697), [#698](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/698)
- Reverse for AbstractGridArrays [#691](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/691)
- LandGeometry, LandThermodynamics as component of LandModel, DryLandModel [#677](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/677)
- Diagnostic albedo separate for ocean/land [#677](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/677)
- Increasing precision for accumulated precipitation output [#685](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/685)
- Allow steps to be specified for for run!(simulation, steps=n) [#684](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/684)
- Restrict extended CI tests to Julia v1.10 (due to Enzyme) [#681](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/681)
- SigmaCoordinates constructors [#679](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/679)
- Full timestepping differentiable by Enzyme with Julia 1.10 [#656](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/656)
- Type instabilities etc flagged by Reactant [#674](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/674)
- Added JLD2 Output to directly output PrognosticVariables and DiagnosticVariables [#675](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/675)
- Compat for GPUArrays v11, JLArrays 0.2, KernelAbstractions 0.9 [#658](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/658)
- MITgcm's 2-layer land bucket model [#672](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/672)
- Fix grid_cell_average! bugs for some grids and resolutions [#673](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/673)
- Interfaces for interpolation of AbstractGridArray [#671](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/671)
- Test folder sorted into subfolders [#671](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/671)
- Land model modularised + land netCDF output [#671](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/671)

## v0.14

- CompatHelper: Allow for Makie.jl v0.22 [#663](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/663)
- Prescribed ocean/land heat fluxes modular [#667](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/667)
- NetCDF output for tracers defined, precipitation rate output at initialize! again [#657](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/657)
- Precipitation rate from large-scale condensation/convection for coupling [#669](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/669)
- Fixed a bug, so that sum now works for LowerTraingularArrays [#668](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/668)
- Require only AbstractVectors in the interpolation [#665](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/665)
- Intitialize! simulation restructured [#666](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/666)
- First time steps in main time loop [#659](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/659)
- Generalise OneBandRadiation to NBandRadiation [#633](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/633)
- New output variables definition simplified [#653](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/653)
- Bug: get_vertices for full grids had southpole at 90˚N [#654](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/654)
- Group initialize!, timestep!, finalize! of main time loop [#652](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/652)
- Coordinates defined in degrees lond, latd, not colat, lon [#647](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/647)
- Spectral Gradients are now differentiable, with correctness check in extended CI [#638](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/638)
- set! for time step [#650](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/650)
- Bug: ZonalWind had sqrt of negative for uncommon resolutions [#649](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/649)
- Tracer advection [#579](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/579)
- Buildkite CI with dummy pipeline [#646](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/646)
- ConvectiveHeating implemented [#639](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/639)
- Number format flexibility with set! [#634](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/634)
- Forcing/drag for primitive models [#635](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/635)
- RingGrids indexing with leading Colon should now always return another RingGrid instance [#637](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/637)
- Roll back GPUArrays upgrade to ensure CUDA compatibility [#636](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/636)
- Change default timestep to 40min at T31 [#623](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/623)

## v0.13

- AbstractSurfacePerturbation introduced [#631](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/631)
- AbstractStochasticPhysics and SPPT implemented [#629](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/629)
- Introduced seperate extended tests that are not run on every commit [#628](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/628)
- One-band longwave radiation [#624](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/624)
- compat entry for FiniteDifferences.jl [#620](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/620)
- Slab ocean and net surface fluxes [#613](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/613)

src/parameters.jl
"""
Parameter struct that holds all parameters that define the default model setup.
With keywords such that default values can be changed at creation.
"""
@with_kw struct Parameters

    # NUMBER FORMATS
    NF::DataType=Float64    # number format

    # RESOLUTION
    trunc::Int=31                      # spectral truncation
    nlon::Int=roundup_fft(3*trunc+1)   # number of longitudes 
    nlat::Int=nlon÷2                   # number of latitudes
    
    nlev::Int=8             # number of vertical levels
    ntracers::Int=1         # number of tracers (specific humidity is one)

    # PHYSICAL CONSTANTS
    R_earth::Real=6.371e6   # Radius of Earth [m]
    Ω::Real=7.292e-5        # Angular frequency of Earth's rotation [1/s]
    gravity::Real=9.81      # Gravitational acceleration [m/s^2]
    akap::Real=2/7          # Ratio of gas constant to specific heat of dry air
                            # at constant pressure = 1 - 1/γ where γ is the
                            # heat capacity ratio of a perfect diatomic gas (7/5)
    cp::Real=1004.0         # Specific heat at constant pressure [J/K/kg]
    R_gas::Real=akap*cp     # Specific gas constant for dry air [J/kg/K]
    alhc::Real=2501.0       # Latent heat of condensation [J/g] for consistency with
                            # specific humidity [g/Kg]
    alhs::Real=2801.0       # Latent heat of sublimation [?]
    sbc::Real=5.67e-8       # Stefan-Boltzmann constant [W/m^2/K^4]

    # STANDARD ATMOSPHERE
    lapse_rate::Real=6.0    # Reference temperature lapse rate -dT/dz [K/km]
    temp_ref::Real=288      # Reference absolute temperature at surface z=0 [K]
    temp_top::Real=216      # Reference absolute temperature in stratosphere [K]
    scale_height::Real=7.5  # Reference scale height for pressure [km]
    pres_ref::Real=1013     # Reference surface pressure [hPa]
    scale_height_humid::Real=2.5    # Reference scale height for specific humidity [km]
    relhumid_ref::Real=0.7          # Reference relative humidity of near-surface air [1]
    water_pres_ref::Real=17         # Reference saturation water vapour pressure [Pa]

    # VERTICAL COORDINATES
    # of the nlev vertical levels, defined by a generalised logistic function,
    # interpolating ECMWF's L31 configuration
    GLcoefs::GenLogisticCoefs=GenLogisticCoefs()
    n_stratosphere_levels::Int=2    # number of vertical levels used for the stratosphere

    # DIFFUSION AND DRAG
    diffusion_power::Real=4                 # Power n of Laplacian in horizontal diffusion ∇²ⁿ
    diffusion_time::Real=2.4                # Diffusion time scale [hrs] for temperature and vorticity
    diffusion_time_div::Real=diffusion_time # Diffusion time scale [hrs] for divergence           
    diffusion_time_strat::Real=12           # Diffusion time scale [hrs] for extra ∇² in the stratosphere
    damping_time_strat::Real=24*30          # Damping time [hrs] for drag on zonal-mean wind in the stratosphere     

    # PARAMETRIZATIONS
    seasonal_cycle::Bool=true   # Seasonal cycle?
    n_shortwave::Int=3          # Compute shortwave radiation every n steps
    sppt_on::Bool=false         # Turn on SPPT?

    # TIME STEPPING
    nstepsday::Int=36           # number of time steps in one day
    Δt::Real=86400/nstepsday    # time step in seconds
    ndays::Real=10              # number of days to integrate for

    # NUMERICS
    robert_filter::Real=0.05    # Robert (1966) time filter coefficeint for suppress comput. mode
    williams_filter::Real=0.53  # William's time filter (Amezcua 2011) coefficient for 3rd order acc
    α::Real=0.5                 # coefficient for semi-implicit computations
                                # 0 -> forward step for gravity wave terms,
                                # 1 -> backward implicit
                                # 0.5 -> centered implicit
    recompute_legendre::Bool=false

    # BOUNDARY FILES
    boundary_path::String=""    # package location is default
    orography_path::String=boundary_path
    orography_file::String="orography_F512.nc"

    # INITIAL CONDITIONS
    initial_conditions::Symbol=:rest    # :rest or :restart

    # OUTPUT
    verbose::Bool=true          # print dialog for feedback
    output::Bool=false          # Store data in netCDF?
    output_dt::Real=6           # output time step [hours]
    output_startdate::DateTime=DateTime(2000,1,1)
    outpath::String=pwd()       # path to output folder
    output_vars::Vector{String}=["u","v","T","humid","logp0"]
    compression_level::Int=3    # 1=low but fast, 9=high but slow
    keepbits::Int=10            # mantissa bits to keep for every variable 

    # TODO assert not allowed parameter values
    @assert α in [0,0.5,1] "Only semi-implicit α = 0, 0.5 or 1 allowed."
end